### PR TITLE
fix(iam): failed to delete a user who is an archived project's Owner or DBA

### LINKED ISSUE #2284: [Bug]: If the user joins a project that has been archived, deleting the user fails
### ODC version

ODC 4.2.4-bp1

### OB version

-

### What happened?

If the user joins a project that has been archived, deleting the user fails.
PS: Only when the user is the administrator/DBA of the project will it not be deleted.

### What did you expect to happen?

The user can be successfully deleted.

### How can we reproduce it (as minimally and precisely as possible)?

1. Add a user to a project
2. Archive the project
3. Delete the user

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by yiminpeng**
#### What type of PR is this?
type-bug

#### What this PR does / why we need it:
If the user joins a project that has been archived, deleting the user fails.
PS: Only when the user is the admin/DBA of the project will it not be deleted.

#### Which issue(s) this PR fixes:

Fixes #2284 

#### Additional documentation e.g., usage docs, etc.:

### CODE REVIEW

**Review by MarkPotato777 — COMMENTED** [review]
Congratulations on submitting your first PR!

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java [hunk 1]
```diff
@@ -142,18 +144,22 @@ private void approvalFlowConfigUsageCheck(Long flowConfigId) {
...
+                        .filter(e -> id2Project.containsKey(e.getKey()))
                         .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
         if (projectId2ResourceRoleNames.size() == 0) {
             return;
         }
-        Map<Long, ProjectEntity> id2Project =
-                projectRepository.findAllByOrganizationId(organizationId).stream().filter(p -> p.getArchived() == false)
-                        .collect(Collectors.toMap(ProjectEntity::getId, p -> p));
-        String names = projectId2ResourceRoleNames.keySet().stream().map(id -> id2Project.get(id).getName())
+        String names = projectId2ResourceRoleNames.keySet().stream()
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java:160]: As you have already filtered by the project id, you could always assume the `id2Project.get(id)` will never return `null` useless JDK has bugs :), even if checking for null is a good programming practice~

**Review by MarkPotato777 — COMMENTED** [review]
PR title suggestion: fix(iam): failed to delete a user who is an archived project's Owner or DBA

File: .idea/runConfigurations/mvn_clean_install_test.xml [hunk 2]
```diff
@@ -26,7 +26,7 @@
           <option name="environmentProperties">
             <map />
           </option>
-          <option name="jreName" value="#USE_PROJECT_JDK" />
+          <option name="jreName" value="1.8" />
```
[the remark this answers is not in the document]
Comment (yiminpeng) [reply .idea/runConfigurations/mvn_clean_install_test.xml:29]: No, I didn't.

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java [hunk 3]
```diff
@@ -142,18 +144,22 @@ private void approvalFlowConfigUsageCheck(Long flowConfigId) {
...
+                .map(id -> Optional.ofNullable(id2Project.get(id))
+                        .map(ProjectEntity::getName).orElseThrow(() -> new UnexpectedException("空指针异常！")))
```
[the remark this answers is not in the document]
Comment (yiminpeng) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java:162]: Got it >_o

File: .idea/runConfigurations/mvn_clean_install_test.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply .idea/runConfigurations/mvn_clean_install_test.xml:29]: Then please do not modify this file if not necessary

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java [hunk 4]
```diff
@@ -142,18 +144,22 @@ private void approvalFlowConfigUsageCheck(Long flowConfigId) {
...
         if (projectId2ResourceRoleNames.size() == 0) {
             return;
         }
-        Map<Long, ProjectEntity> id2Project =
-                projectRepository.findAllByOrganizationId(organizationId).stream().filter(p -> p.getArchived() == false)
-                        .collect(Collectors.toMap(ProjectEntity::getId, p -> p));
-        String names = projectId2ResourceRoleNames.keySet().stream().map(id -> id2Project.get(id).getName())
+        String names = projectId2ResourceRoleNames.keySet().stream()
+                .map(id -> Optional.ofNullable(id2Project.get(id))
+                        .map(ProjectEntity::getName).orElseThrow(() -> new UnexpectedException("NullPointerException")))
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java:162]: The exception error message is supposed to be readable, but `NullPointerException` does not bring any useful information.
Message suggestion: "project not found, project id=xxx". Or as I said in the previous comment, you could always assume you would never get `null` as you have already filtered out these projects, you may just delete these new codes.

Comment (yiminpeng) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java]: got it

**Review by MarkPotato777 — APPROVED** [review]
LGTM

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**